### ISSUE TITLE
UnicodeDecodeError on Python 3

### ISSUE THREAD

**Issue by codekoala** [fgimian/easysnmp#8]
The following code works with Python 2.7.10:

```
from easysnmp import Session
sess = Session(hostname='localhost', community='public', version=2)
sess.walk('.1.3.6.1.2.1')
```

This same code fails with Python 3.4.3 with the following error:

```
/usr/lib/python3.4/site-packages/easysnmp/session.py in walk(self, oids)
    446 
    447         # Perform the SNMP walk using GETNEXT operations
--> 448         interface.walk(self, varlist)
    449 
    450         # Validate the variable list returned

UnicodeDecodeError: 'utf-8' codec can't decode byte 0x80 in position 0: invalid start byte
```

**Comment by codekoala** [comment]
I've tried out a few different OIDs to see what exactly might be causing the failure. Two of my systems fail when walking ``.1.3.6.1.2.1.2.2.1.6`` with Python 3. Here's what's returned by easysnmp under Python 2:

System 1:
```
[<SNMPVariable value='' (oid='ifPhysAddress', oid_index='1', snmp_type='OCTETSTR')>,
 <SNMPVariable value='$ (contains binary)' (oid='ifPhysAddress', oid_index='2', snmp_type='OCTETSTR')>,
 <SNMPVariable value='$ (contains binary)' (oid='ifPhysAddress', oid_index='3', snmp_type='OCTETSTR')>,
 <SNMPVariable value='$ (contains binary)' (oid='ifPhysAddress', oid_index='4', snmp_type='OCTETSTR')>,
 <SNMPVariable value='* (contains binary)' (oid='ifPhysAddress', oid_index='6', snmp_type='OCTETSTR')>,
 <SNMPVariable value='$ (contains binary)' (oid='ifPhysAddress', oid_index='7', snmp_type='OCTETSTR')>]
```

System 2:
```
[<SNMPVariable value='' (oid='ifPhysAddress', oid_index='1', snmp_type='OCTETSTR')>,
 <SNMPVariable value='r? (contains binary)' (oid='ifPhysAddress', oid_index='2', snmp_type='OCTETSTR')>,
 <SNMPVariable value='r? (contains binary)' (oid='ifPhysAddress', oid_index='3', snmp_type='OCTETSTR')>,
 <SNMPVariable value='Vz (contains binary)' (oid='ifPhysAddress', oid_index='4', snmp_type='OCTETSTR')>,
 <SNMPVariable value='' (oid='ifPhysAddress', oid_index='10', snmp_type='OCTETSTR')>,
 <SNMPVariable value='r? (contains binary)' (oid='ifPhysAddress', oid_index='58', snmp_type='OCTETSTR')>]
```

**Comment by fgimian** [comment]
Thanks for the report.  I can definitely reproduce it here too.  I'll attempt to get this fixed for the 0.2.2 release.  :smile: 

Cheers
Fotis

**Comment by fgimian** [comment]
Ok, I've spent some time on this tonight and know what's going on.  Python 3.x as we know uses UTF-8 encoding for strings by default and this too was the case with retrieved values via SNMP when we build our value object:

```c
        PyObject* val_obj =  (val ?
                              Py_BuildValue("s#", val, len) :
                              Py_BuildValue(""));
```

Though, a lot of data provided via SNMP can actually contain binary data or technically, data in any encoding which we're not yet aware of.  I have a working version now which corrects this by simply using bytes instead of unicode strings as returns, but I really would like to think through the best way to apply this strategy across the codebase.

Another option is to encode the string in latin-1 encoding, which is pretty much the same as binary.

Honestly, I think that Python 3.x has complicated the world for us all.  As much as I appreciate the need for UTF-8, making it the default encoding used for all strings is arguably a poor design decision.  Oh well, we have to live with this I guess :smile: 

Leave ti with me, I'd like to come up with a good solution :smile: 
Fotis

**Comment by fgimian** [comment]
Hey there, can you please try again with this branch: https://github.com/fgimian/easysnmp/tree/better-python3

I've re-implemented handling of all strings to be latin-1 encoded unicode across Python 2.x and 3.x.  In a future option, I'll allow the user to specify an encoding of their choice.

Cheers
Fotis

**Comment by codekoala** [comment]
Sorry for the delay. It seems the branch has been merged into master and that you've cut a 0.2.2 release. I've tested this on both Python 2.7.10 and 3.4.3 on my local system, and it's working exactly the same now.

Thank you!!!

**Comment by fgimian** [comment]
Awesome! My pleasure, I'm just waiting for the Travis build to pass and I'll upload 0.2.2 to PyPi :smile: 

Cheers
Fotis
